### PR TITLE
feat: add EXLA.to_mlir_module/2

### PR DESCRIPTION
**Pull request by polvalente**
- feat: add EXLA.to_mlir_module/3
- doc: add doctest

### CODE REVIEW

File: exla/lib/exla/mlir/module.ex [hunk 1]
```diff
@@ -71,6 +71,10 @@ defmodule EXLA.MLIR.Module do
     * `:use_spmd` - enables Single-Program Multiple-Data partioning.
       This is set to true if `:num_partitions` is more than one, otherwise is `false`.
 
+    * `:module_compilation` - either `:to_mlir` or `:to_pjrt`. The default is `:to_pjrt`.
```
Comment (polvalente) [inline exla/lib/exla/mlir/module.ex:74]: @josevalim not sure about the option naming here

Comment (josevalim) [reply exla/lib/exla/mlir/module.ex:74]: It is private, so it is fine!

File: exla/lib/exla.ex [hunk 2]
```diff
@@ -355,6 +355,43 @@ defmodule EXLA do
...
+          %2 = stablehlo.add %0, %1 : tensor<f32>
+          stablehlo.return %2 : tensor<f32>
+        }
+      }
+      \"\"\"
+  """
+  def to_mlir_module(function, args, options \\ []) do
+    comp_fun = fn _key, callback ->
+      {:ok, {_xla_time, executable, _extra, _outfeed}} = callback.()
+      throw({:mlir_module, executable.ref})
```
Comment (josevalim) [inline exla/lib/exla.ex:382]: Gah, this is relying on internals of another module but I can't think of anything better for now, so ship it.

**Review by josevalim — APPROVED** [review]

File: exla/lib/exla/mlir/module.ex (hunk 1 above)
Comment (jonatanklosko) [reply exla/lib/exla/mlir/module.ex:74]: If we return `%EXLA.MLIR.Module{}` in the other function, shouldn't the whole module be public?

Comment (josevalim) [reply exla/lib/exla/mlir/module.ex:74]: Good catch. We should return the string only, not the module struct.

File: exla/lib/exla.ex [hunk 3]
```diff
@@ -355,6 +355,43 @@ defmodule EXLA do
...
+  Takes in a function, the templates variables and the compilation options
+  and returns the `EXLA.Executable` struct.
+
+  ## Examples
+
+      iex> fun = fn x, y -> Nx.add(Nx.sin(x), Nx.cos(y)) end
+      iex> args = [1.0, 2.0]
+      iex> module = EXLA.to_mlir_module(fun, args)
+      iex> EXLA.MLIR.Module.as_string(module)
+      ~c\"\"\"
```
Comment (jonatanklosko) [inline exla/lib/exla.ex:368]: Btw. let's make `Module.as_string` return an actual string, not charlist :D

Comment (polvalente) [reply exla/lib/exla.ex]: Done

File: exla/lib/exla/mlir/module.ex (hunk 1 above)
Comment (polvalente) [reply exla/lib/exla/mlir/module.ex:74]: The public-facing function is now returning the binary directly

**Pull request merged**